### PR TITLE
`lsp-rename`: improve UI

### LINKED ISSUE #2314: `lsp-rename`: check if supported before prompting for a name
When I start using a new language sever, I don't know what it supports. If because of that I try to rename something, even though the language server doesn't support it, I just end up wasting my time coming up with a new name.

`lsp-rename` should do a capability check + prepareRename before asking the user for a new name. This will improve discoverability, because users aren't punished for not knowing that rename doesn't work.

This will be an easy fix.

### PR DESCRIPTION
**Pull request by nbfalcon**
`lsp-rename` doesn't check whether renaming is supported first. Do that,
throwing an error otherwise.

Fixes #2314.

### CODE REVIEW

**Comment by yyoncho** [comment]
I am not sure we need anything more than...

```
(unless (lsp-feature? "textDocument/rename")
  (user-error "Rename not..." ))
```

in the `lsp-rename` interactive form.

**Comment by nbfalcon** [comment]
Don't merge yet - there is a bug that I need to fix. As to your comment, if we use that new function, there will be a nice LSP :: message instead. Do you consider that undesirable? If so, I'll revert.

**Comment by nbfalcon** [comment]
Also, the prompt is now a bit nicer, because I don't use buffer-substring-no-properties.

**Comment by nbfalcon** [comment]
- The bug is fixed; rename now works
- I use `lsp-feature?` now, thanks for the tip. It is really much simpler, and probably less fragile.

File: lsp-mode.el
Comment (yyoncho) [inline lsp-mode.el:5665]:
```
(-if-let (lsp-feature? "xxx")
    lsp-feature?
  ...)
```

returns "xxx".

Can you use the unless form as suggested?

File: lsp-mode.el
Comment (yyoncho) [inline lsp-mode.el:5682]: can you avoid putting comments like that? This should be evident from the content of the next line.

Comment (nbfalcon) [reply lsp-mode.el]: I usually don't do that - this one was an accident. Fixed it.

**Comment by nbfalcon** [comment]
Note that the error messages aren't propertized, which may require further thought.

File: lsp-mode.el
Comment (yyoncho) [inline lsp-mode.el:5680]: 1. prepareRename does not work with `lsp-feature?`
2. Can you revert the renamed symbols?
3. Can you remove irrelevant comments?

File: lsp-mode.el
Comment (yyoncho) [inline lsp-mode.el:1191]: seems like this is redundant: there is lsp--error.

**Comment by nbfalcon** [comment]
Because of the unpropertized error messages, I'd advise you to wait before merging.

File: lsp-mode.el
Comment (yyoncho) [inline lsp-mode.el:1198]: I don't see any effect from this propertize - the message is displayed with default font.

File: lsp-mode.el
Comment (yyoncho) [reply lsp-mode.el:1191]: Ignore... I missed the fact that it is using error not message. But still, is this function used in this PR? Are we going to replace all error calls with that?

File: lsp-mode.el
Comment (nbfalcon) [reply lsp-mode.el:1198]: Yes, that is the issue currently.

File: lsp-mode.el
Comment (nbfalcon) [reply lsp-mode.el:1191]: `lsp--error` logs a `message`, which doesn't cause evaluation to abort.

Comment (nbfalcon) [reply lsp-mode.el:1191]: This function is used by this PR, which is why I added it. However, replacing all error calls might be a bit much. From looking at the code, the `(error (lsp-error ...))` idiom seems common, however it doesn't solve the text-properties issue.

File: lsp-mode.el
Comment (nbfalcon) [reply lsp-mode.el:5680]: 1. What should I use instead? The old code?
2, 3. Will do.

Comment (yyoncho) [reply lsp-mode.el]: The old code should be fine.

Comment (nbfalcon) [reply lsp-mode.el]: Are you sure that it doesn't though? From stepping trough that with edebug, `lsp-feature?` returned a complex result for "textDocument/prepareRename", while consistently returning nil if I evaluate it somewhere without `lsp-mode` enabled (a *scratch* buffer, for instance).

Comment (yyoncho) [reply lsp-mode.el]: I checked and  lsp-method-requirements are actually implemented so lsp-feature? will work. It is odd that we haven't replaced the `lsp--get-symbol-to-rename`

**Comment by nbfalcon** [comment]
I cleaned up the history a bit. Aside from that, I eliminated the new error functions, since they don't work they way I wanted anyway, and use `user-error' and normal `error' instead. This is probably also more idiomatic.

**Comment by nbfalcon** [comment]
This PR is now ready.

File: lsp-mode.el
Comment (yyoncho) [inline lsp-mode.el:5674]: please revert that.

File: lsp-mode.el
Comment (yyoncho) [inline lsp-mode.el:5665]: why do you rename symbol? Why do you introduce when-let? 

Let's clarify that for future reference - do not introduce random renames, restructuring, etc - nobody is doing this PR in this repo. You are making the life of the reviewers harder than it should be. I have to go and test each that you have changed to verify that each line of all lines that you have changed does not introduce a regression. I cannot spend that much time reviewing trivial changes chasing for regressions.

File: lsp-mode.el
Comment (kiennq) [inline lsp-mode.el:5662]: Why we need to use `buffer-substring` instead of `-no-properties` version?
Do we need the highlighting when prompting?

File: lsp-mode.el
Comment (kiennq) [inline lsp-mode.el:5668]: Why don't we just check the capability in this command (public) instead of changing internal one?
I don't think guarding every internal one is desirable/good from perf POV. Just guard once in public one should be enough, shouldn't it?

Comment (nbfalcon) [reply lsp-mode.el:5668]: If we guard only in `lsp-rename`, the user would be prompted to enter a new name even though renaming is unsupported.

File: lsp-mode.el
Comment (nbfalcon) [reply lsp-mode.el:5662]: Compare:

VS with highlighting:


Do you prefer the former?

File: lsp-mode.el
Comment (nbfalcon) [reply lsp-mode.el:5668]: Also, it's fast enough:
```emacs-lisp
(with-current-buffer "main.c"
  (benchmark-run 1000 (ignore (lsp-feature? "textDocument/rename"))))
```
yields 0.037 seconds.

Calling it only twice gives me 2ms timings, which is less than a 120FPS monitor's refresh latency.

File: lsp-mode.el
Comment (nbfalcon) [reply lsp-mode.el]: Ok, I want to do that in the future. The `when-let` is because that function should return nil if there is nothing to rename at point. Without that `when-let`, if `(thing-at-point symbol)` were to yield nil, the function would return `(cons nil nil)`, instead of nil as it should. This, unlike I had assumed, didn't cause any real problems, because -if-let checks the bindings, not the input against nil.

```emacs-lisp
(eq nil nil) => true
(eq nil (cons nil nil)) => false
(equal nil (cons nil nil)) => false
```

File: lsp-mode.el
Comment (yyoncho) [reply lsp-mode.el:5668]: > If we guard only in `lsp-rename`, the user would be prompted to enter a new name even though renaming is unsupported.

see https://github.com/emacs-lsp/lsp-mode/pull/2317#issuecomment-724147973
https://github.com/yyoncho/lsp-mode/commit/443f801a4117361d2247c408b02dcc0d8c95a441 .

Comment (nbfalcon) [reply lsp-mode.el:5668]: Is it better to move the check outside of `lsp--get-symbol-to-rename` though?

File: lsp-mode.el
Comment (yyoncho) [reply lsp-mode.el]: You will still rename all local variables, introduce random changes without documenting/testing them based on assumptions and you will expect the rest of the team to do that for you?!

Comment (nbfalcon) [reply lsp-mode.el]: Sorry, I missspelled won't...

File: lsp-mode.el
Comment (nbfalcon) [reply lsp-mode.el:5668]: It's actually called only once though, in `lsp--get-symbol-to-rename`. We needn't add checks in every function, because the server will respond with a proper "not implemented" error if it isn't supported. Here, we cannot rely on that though, since we mustn't prompt the user for a name if the rename wouldn't work anyway.

**Comment by nbfalcon** [comment]
I'll extend this to highlight the symbol being renamed; with the refactoring necessary do that, I can add tests ensuring that it will work in all cases.

**Comment by nbfalcon** [comment]
The various `lsp-rename' helper functions are now thoroughly tested, and the symbol being renamed is highlighted.

**Comment by nbfalcon** [comment]
In summary:
- `lsp-rename` now checks whether it is supported in the interactive form, trough a new function
- `lsp-rename` no longer user errors if the input is nil
- Renamed identifiers are marked in the buffer using `lsp-face-rename`
- The symbol to be renamed is propertized in the prompt, as is the initial input
- Lots of tests for the two functions. I'll need to investigate the CI failures; the tests work in isolation on my machine just fine.

**Comment by nbfalcon** [comment]
Here is a screenshot:

File: lsp-mode.el
Comment (kiennq) [reply lsp-mode.el:5717]: I think you don't get my point. Let say if the `lsp--get-symbol-to-rename` is gone/refactored, which can happen since it's internal function, now we will need to find a new place to put the capacity checking.
On other hand, `lsp-rename` is public function, and is unlikely to change without a breaking change, put the capacity guard there is the best, since user has to use that function to actually go through rename.
Is there any specific concern why the capacity guard has to be put in `lsp--get-symbol-to-rename`?

File: lsp-mode.el
Comment (kiennq) [inline lsp-mode.el:5691]: It would be better to specify what the structure look like instead of just saying it will be the return value from `lsp--get-symbol-to-rename`

File: lsp-mode.el
Comment (kiennq) [inline lsp-mode.el:5711]: nit: is there a need for `-doto` in this case, since there's only on form?

File: lsp-mode.el
Comment (nbfalcon) [reply lsp-mode.el:5691]: The structure is defined in `lsp--get-symbol-to-rename`'s docstring.

File: lsp-mode.el
Comment (nbfalcon) [reply lsp-mode.el:5717]: We could add an additional `lsp-feature?` check to `lsp-rename`. However, the check in `lsp--get-symbol-to-rename` must stay: the interactive block is executed before the body, meaning the user would get prompted for the symbol to rename even though rename might not be possible at all or at point.

File: lsp-mode.el
Comment (nbfalcon) [reply lsp-mode.el]: Fixed, thanks.

**Comment by nbfalcon** [comment]
There are two issues with this as of now (hence the "Draft" status):
- [ ] `lsp-rename` sometimes override symbol-highlight overlays, and sometimes don't, which leads to incosistency.
- [ ] For some language servers, `prepareRename` can take too long or time out (`rust-analyzer`). Maybe somehow waiting until C-g would be better here, or perhaps setting a lower timeout, notifying the user somehow. Not using `prepareRename` is also a possibility, but is certainly sub-optimal.

And perhaps some ToDos:
- [ ] Support rename preview, like VSCode

**Comment by danielmartin** [comment]
> And perhaps some ToDos:
> 
>   Support rename preview, like VSCode

I'm working on this feature at the moment.  It'll be orthogonal to the changes proposed here (if accepted).  You'll enter the rename preview mode if you invoke `lsp-rename` with `C-u`.

File: lsp-mode.el
Comment (kiennq) [reply lsp-mode.el:5732]: You can put the check in `(interactive)` form too. Like what @yyoncho suggested.
If you hate that, you can even make `(interactive '(nil))` and check for `newName` outside of `interactive` block and call to `lsp--get-symbol-to-rename`.

File: lsp-mode.el
Comment (yyoncho) [inline lsp-mode.el:5698]: Shouldn't this variable be public to avoid persisting it?

File: lsp-mode.el
Comment (yyoncho) [inline lsp-mode.el:5718]: I would vote for using different faces for placeholder and for the thing to rename. Using `bold` face seems to be a good candidate for one of them (to avoid being too colorful).

File: lsp-mode.el
Comment (yyoncho) [inline lsp-mode.el:5723]: here we get an error if you kill the buffer which contains the thing to rename and then press ret in the read-string.

**Review by yyoncho — COMMENTED** [review]
Looks good. Please drop a note which servers you have tested with. 

Few minor nits suggestions:

File: lsp-mode.el
Comment (nbfalcon) [reply lsp-mode.el:5698]: You make a good point, and from looking at other packages (`magit`, `evil (evil-ex)`), they also seem to be using public history variables. I'm making this public.

File: lsp-mode.el
Comment (nbfalcon) [reply lsp-mode.el:5718]: Do you propose overriding the face of the thing to be renamed the same way as is done for the placeholder? What we might also want to explore is adding a face to it, using `add-face-text-property`. I'll post some screenshots of various approaches soon.

**Comment by nbfalcon** [comment]
I tested with `jdtls`, `rust-analyzer`, `ccls` and `pyright`.

File: lsp-mode.el
Comment (nbfalcon) [reply lsp-mode.el:5723]: That is unavoidable, because `read-string` errors if the old buffer was killed. However, I don't see how that would cause issues? Overlays can be deleted if the old buffer is dead; the error would just get propagated. I think this should be fixed in Emacs core, because all packages can't be expected to guard against a buffer-live-p error just to be able to use `read-string`.

`delete-overlay` + dead-buffer:
```elisp
(delete-overlay
 (with-temp-buffer
   (insert "foo")
   (make-overlay 0 1)))
```

Bind `M-k` to kill the old buffer in the minibuffer map:
```elisp
(define-key minibuffer-local-map (kbd "M-k") (lambda () (interactive) (kill-buffer (window-buffer (minibuffer-selected-window)))))
```

**Comment by nbfalcon** [comment]
`rust-analyzer` seems to send incorrect ranges in certain cases (sometimes with an extra :uri), which causes the rename overlays and placeholder to be borked (not a regression, master also has this). This is a spec violation though, as the result shall be a Range, not something containing a :uri.

**Comment by nbfalcon** [comment]
Digging trough RA's code, it seems that the strange results are fixed, at least in master, because there is no code returning a uri in `handle_prepare_rename`.

File: lsp-mode.el
Comment (yyoncho) [reply lsp-mode.el:5723]: Ok - I thought the error can be prevented.

Comment (yyoncho) [reply lsp-mode.el:5723]: .. easily.

File: lsp-mode.el
Comment (yyoncho) [reply lsp-mode.el:5718]: I am proposing using 2 different faces like this:

``` emacs-lisp
(let ((placeholder "foo")
      (rename-me "foo"))
  (read-string (format "Rename %s to: " (propertize rename-me 'face 'bold)) 
               (propertize placeholder 'face 'font-lock-variable-name-face)
               'lsp-rename-history))
```

**Review by yyoncho — APPROVED** [review]

**Comment by yyoncho** [comment]
Thank you.

I merged it as it is - the cosmetics about the faces can be discussed separatelly.

**Pull request merged**
